File: frontend/mobile/lib/features/auth/data/auth_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/config/env.dart';
import '../../../core/network/api_client.dart';
import '../domain/auth_user.dart';

abstract class AuthRepository {
  Future<({String accessToken, AuthUser user})> login(String email, String password);
  Future<AuthUser> me();
}

class _HttpAuthRepository implements AuthRepository {
  _HttpAuthRepository(this._dio);
  final Dio _dio;

  @override
  Future<({String accessToken, AuthUser user})> login(String email, String password) async {
    final r = await _dio.post('/auth/login', data: {'email': email, 'password': password});
    final token = r.data['access_token'] as String;
    final meResp = await _dio.get(
      '/profile/me',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return (accessToken: token, user: AuthUser.fromJson(meResp.data as Map<String, dynamic>));
  }

  @override
  Future<AuthUser> me() async {
    final r = await _dio.get('/profile/me');
    return AuthUser.fromJson(r.data as Map<String, dynamic>);
  }
}

class _MockAuthRepository implements AuthRepository {
  @override
  Future<({String accessToken, AuthUser user})> login(String email, String password) async {
    await Future.delayed(const Duration(milliseconds: 400));
    if (password.length < 4) {
      throw DioException(
        requestOptions: RequestOptions(path: '/auth/login'),
        response: Response(
          requestOptions: RequestOptions(path: '/auth/login'),
          statusCode: 401,
          data: {'message': 'invalid credentials'},
        ),
        message: 'invalid credentials',
      );
    }
    return (
      accessToken: 'mock-token',
      user: AuthUser(id: 'demo', email: email, fullName: 'Demo Patient', role: 'patient'),
    );
  }

  @override
  Future<AuthUser> me() async {
    await Future.delayed(const Duration(milliseconds: 200));
    return AuthUser(id: 'demo', email: 'demo@medapp.test', fullName: 'Demo Patient', role: 'patient');
  }
}

final authRepositoryProvider = Provider<AuthRepository>((ref) {
  if (Env.useMockApi) return _MockAuthRepository();
  return _HttpAuthRepository(ref.watch(dioProvider));
});

File: frontend/mobile/src/features/auth/authRepository.ts
import { AxiosError } from "axios";
import { apiClient } from "@/core/network/apiClient";
import { env } from "@/core/config/env";
import {
  type AuthApiUser,
  type AuthUser,
  type LoginResult,
  toAuthUser,
} from "@/features/auth/types";

export interface AuthRepository {
  login(email: string, password: string): Promise<LoginResult>;
  me(): Promise<AuthUser>;
}

class HttpAuthRepository implements AuthRepository {
  async login(email: string, password: string): Promise<LoginResult> {
    const loginRes = await apiClient.post<{ access_token: string }>("/auth/login", {
      email,
      password,
    });
    const accessToken = loginRes.data.access_token;

    const meRes = await apiClient.get<AuthApiUser>("/profile/me", {
      headers: { Authorization: `Bearer ${accessToken}` },
    });
    return { accessToken, user: toAuthUser(meRes.data) };
  }

  async me(): Promise<AuthUser> {
    const res = await apiClient.get<AuthApiUser>("/profile/me");
    return toAuthUser(res.data);
  }
}

class MockAuthRepository implements AuthRepository {
  async login(email: string, password: string): Promise<LoginResult> {
    await delay(400);
    if (password.length < 4) {
      throw new AxiosError(
        "invalid credentials",
        "ERR_BAD_REQUEST",
        undefined,
        undefined,
        {
          status: 401,
          statusText: "Unauthorized",
          data: { message: "invalid credentials" },
          headers: {},
          config: {} as never,
        }
      );
    }
    return {
      accessToken: "mock-token",
      user: {
        id: "demo",
        email,
        fullName: "Demo Patient",
        role: "patient",
      },
    };
  }

  async me(): Promise<AuthUser> {
    await delay(200);
    return {
      id: "demo",
      email: "demo@medapp.test",
      fullName: "Demo Patient",
      role: "patient",
    };
  }
}

const delay = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

export const authRepository: AuthRepository = env.useMockApi
  ? new MockAuthRepository()
  : new HttpAuthRepository();
